### DISCUSSION TITLE
Random crashes  vs Swizzling

### DISCUSSION THREAD

**Discussion by choulepoka** [getsentry/sentry-cocoa#3559] [Q&A]
Hello there!

I have a curious story to tell, and perhaps this will be of use to somebody else in the Sentry Customer Community.

I have been tracking a random crash in our app for the last 2 years, the crash occurs when downloading a file locally (i.e.: `URLSessionDownloadTask`). Always seemed to be near the NSProgress update process, but the stack trace keep changing, and there are no references to any of our code.

I always assumed that this was caused by some obscure bug in our code, because the crash occurred randomly, and no matter what I did, what the user were using (device, OS version, app version), there would always be at least 1 or 2 daily occurence of these crashes.

Fast forward December 2023, where due to a bad merge conflict resolution, swizzling was disabled for Sentry, and it went live on the App store. (TLDR, no new build would be uploaded until mid January, where I could reactivate Swizzling)

However, much to my surprise, after almost 4 full weeks on the app store with the new version (with 90% upgrade rate for our user-base), there are zero crash for the last 4 weeks for that specific build. The users that did _not_ upgrade to that specific build continued to experience the crashes, and the newest build that had the swizzling restored also started to exhibit the crashes too.

So I am tempted to permanently disabled Swizzling, even though I would loose some theoretically useful feature, but the fact is: I am only really interested by the crash reports

For anybody out there experiencing those random crashes, try disabling swizzling, it may help.

**Comment by philipphofmann** [comment]
@choulepoka, I'm sorry you experience random crashes when swizzling is enabled. Can you share some of them with us so we have to change to look into them and fix them?

We sometimes think about adding a guide or an extra package for auto-performance tracing that does not require the usage of swizzling. The downside is that it will require manual setup, but the benefit is no swizzling. Would you be interested in such a solution?

**Comment by Petrolion** [comment]
@choulepoka, @philipphofmann 
The issue may be caused by a missing Main thread dispatch in SentryCrashWrapper -> enrichScope, as shown in the following code:

```
#if SENTRY_TARGET_REPLAY_SUPPORTED

    NSArray<UIWindow *> *appWindows = SentryDependencyContainer.sharedInstance.application.windows;
    if ([appWindows count] > 0) {
        UIScreen *appScreen = appWindows.firstObject.screen;
        if (appScreen != nil) {
            [deviceData setValue:@(appScreen.bounds.size.height) forKey:@"screen_height_pixels"];
            [deviceData setValue:@(appScreen.bounds.size.width) forKey:@"screen_width_pixels"];
        }
    }

#endif 
```


This "UIWindow screen" method needs to run on the Main thread.
In my experience, this issue occurs sporadically — for example, when HTTP requests are triggered on a background queue and the response contains an error status code (e.g., 502). Sentry is then invoked via the swizzled method call, attempts to access the Main thread, and subsequently crashes.



**React Native Project** 
- RN
  - @sentry/react-native: 6.10.0
- iOS
  - RNSentry: 6.10.0
  - Sentry/HybridSDK: 8.48.0

**Example stack trace**
```
=================================================================

Main Thread Checker: UI API called on a background thread: -[UIWindow screen]PID: 1461, TID: 637235, Thread name: (none), Queue name: com.facebook.react.NetworkingQueue, 
QoS: 0Backtrace:4   projectname                             
0x0000000100b19718 -[SentryCrashWrapper enrichScope:] + 9405   projectname                             
0x0000000100b30fe4 -[SentryHub scope] + 1446   projectname                             
0x0000000100b39bb8 -[SentryNetworkTracker addTraceWithoutTransactionToTask:] + 727   projectname                             
0x0000000100b39a34 -[SentryNetworkTracker urlSessionTaskResume:] + 13488   projectname                             
0x0000000100b3b9a8 __57+[SentryNetworkTrackingIntegration swizzleURLSessionTask]_block_invoke_2 + 529   projectname                             
0x0000000100986e7c -[RCTHTTPRequestHandler sendRequest:withDelegate:] + 52010  projectname                             
0x000000010098bef4 -[RCTNetworkTask start] + 12411  projectname                             
0x000000010098a1a8 -[RCTNetworking sendRequest:responseType:incrementalUpdates:responseSender:] + 89212  projectname                             
0x000000010098b4e4 __38-[RCTNetworking sendRequest:callback:]_block_invoke_2 + 15613  libdispatch.dylib                   
0x00000001045c0b98 _dispatch_call_block_and_release + 3214  libdispatch.dylib                   
0x00000001045c27bc _dispatch_client_callout + 2015  libdispatch.dylib                   
0x00000001045ca66c _dispatch_lane_serial_drain + 83216  libdispatch.dylib                   
0x00000001045cb408 _dispatch_lane_invoke + 40817  libdispatch.dylib                   
0x00000001045d8404 _dispatch_root_queue_drain_deferred_wlh + 32818  libdispatch.dylib                   
0x00000001045d7a38 _dispatch_workloop_worker_thread + 44419  libsystem_pthread.dylib             
0x00000001f185c934 _pthread_wqthread + 28820  libsystem_pthread.dylib             
0x00000001f18590cc start_wqthread + 8Main Thread Checker: UI API called on a background thread: -[UIWindow screen]PID: 1461, TID: 637235, Thread name: (none), Queue name: com.facebook.react.NetworkingQueue, QoS: 0Backtrace:4   projectname                             
0x0000000100b19718 -[SentryCrashWrapper enrichScope:] + 9405   projectname                             
0x0000000100b30fe4 -[SentryHub scope] + 1446   projectname                             
0x0000000100b39bb8 -[SentryNetworkTracker addTraceWithoutTransactionToTask:] + 727   projectname                             
0x0000000100b39a34 -[SentryNetworkTracker urlSessionTaskResume:] + 13488   projectname                             
0x0000000100b3b9a8 __57+[SentryNetworkTrackingIntegration swizzleURLSessionTask]_block_invoke_2 + 529   projectname                             
0x0000000100986e7c -[RCTHTTPRequestHandler sendRequest:withDelegate:] + 52010  projectname                             
0x000000010098bef4 -[RCTNetworkTask start] + 12411  projectname                             
0x000000010098a1a8 -[RCTNetworking sendRequest:responseType:incrementalUpdates:responseSender:] + 89212  projectname                             
0x000000010098b4e4 __38-[RCTNetworking sendRequest:callback:]_block_invoke_2 + 15613  libdispatch.dylib                   
0x00000001045c0b98 _dispatch_call_block_and_release + 3214  libdispatch.dylib                   
0x00000001045c27bc _dispatch_client_callout + 2015  libdispatch.dylib                   
0x00000001045ca66c _dispatch_lane_serial_drain + 83216  libdispatch.dylib                   
0x00000001045cb408 _dispatch_lane_invoke + 40817  libdispatch.dylib                   
0x00000001045d8404 _dispatch_root_queue_drain_deferred_wlh + 32818  libdispatch.dylib                   
0x00000001045d7a38 _dispatch_workloop_worker_thread + 44419  libsystem_pthread.dylib             
0x00000001f185c934 _pthread_wqthread + 28820  libsystem_pthread.dylib             
0x00000001f18590cc start_wqthread + 8
```

**Reply by philprime** [reply]
Hi @Petrolion, if you see this warning that means that the swizzling is active while the SDK itself is not. When starting the SDK we are creating and enriching a `scope`, which exists for as long as the SDK is active. 

A similar occurence is using other manual methods of `SentrySDK`, e.g. `SentrySDK.capture(error:)` as seen in #4188 

Are you starting the SDK in AppDelegate on Main Thread?

**Reply by Petrolion** [reply]
Hi @philprime, thanks for the quick reply!
Unfortunately, I can’t reproduce the error reliably; however, I do get the impression that it occurs shortly after the SDK is disabled.

I’m still not entirely following your explanation: How do you infer from the calls visible in the stack trace—or from the implementation of the swizzling method up to enrichScope—that the Sentry SDK is no longer active and and that the scope is related to it??

Additional information
- I start the Sentry SDK from the main thread using DispatchQueue.main.async { … }.
- The SDK is toggled on or off based on the user’s consent decision, both at app launch and at runtime via the settings.
As far as I can see, there is no way to undo swizzling once applied. This inevitably leads to a state where the SDK is disabled but the swizzling remains active.

It thus appears possible that the swizzled method allows the request to flow through to enrichScope. If the network call was made from a background thread, the error occurs.

Even if swizzling remains active, enrichScope shouldn’t invoke any UI code without dispatching to the main thread first.

In the code, the window is access like this: "NSArray<UIWindow *> *appWindows = SentryDependencyContainer.sharedInstance.application.windows;"

Inside, the code is dispatched to the main thread with:

```
[SentryDependencyContainer.sharedInstance.dispatchQueueWrapper
    dispatchSyncOnMainQueue:^{
        // …
    }];
```
Shouldn’t there also be a dispatch for the line "UIScreen *appScreen = appWindows.firstObject.screen;"
to ensure that this access also happens on the main thread?

I just want to make sure that none of the app’s network requests can fail because of this.
I don’t want to disable the swizzling itself, since it’s very helpful. 
Do you have any other ideas what I could do here? Or am I perhaps doing something wrong?

**Reply by philprime** [reply]
Thanks for all the details, I'll try to address them individually:

> I’m still not entirely following your explanation: How do you infer from the calls visible in the stack trace—or from the implementation of the swizzling method up to enrichScope—that the Sentry SDK is no longer active and and that the scope is related to it??

I've seen this exact behavior before while maintaining the SDK, therefore I could infer that this might be the same issue as in #4015.

When bootstrapping the SDK, we are creating internal instances of the classes `SentryHub` and `SentryScope`. In the initializer of `SentryHub`, we are calling the `[_crashWrapper enrichScope:_scope]` and as the SDK should be started on main thread, the access to `UIWindow` is fine. Afterwards the scope is enriched and we do not have to do that again.

https://github.com/getsentry/sentry-cocoa/blob/2e790c65299f50859e726bdbaa93dafe178bf2fb/Sources/Sentry/SentryHub.m#L83

When calling `[SentrySDK close]` the SDK is teared down, discarding the internal instance of `SentryHub` and  therefore also discarding the `SentryScope`.

Now when the network tracker is accessing the `[SentryHub scope]` after the SDK already discarded the `_scope`, it creates a new instance, calling `enrichScope` once again:

https://github.com/getsentry/sentry-cocoa/blob/main/Sources/Sentry/SentryHub.m#L587-L602

This SDK was changed a lot over the years, so there was probably a reason it was designed this way, which might not be relevant anymore today.

>   The SDK is toggled on or off based on the user’s consent decision, both at app launch and at runtime via the settings.
> - As far as I can see, there is no way to undo swizzling once applied. This inevitably leads to a state where the SDK is disabled but the swizzling remains active.
> It thus appears possible that the swizzled method allows the request to flow through to enrichScope. If the network call was made from a background thread, the error occurs.

You are correct with the limited existence/non-existent unswizzling. Swizzling and unswizzling brings high risk, therefore we are always hesitant in changing it. But we added it to the list of changes we might want to tackle with the next major release of the SDK ([see discussion](https://github.com/getsentry/sentry-cocoa/discussions/2946#discussioncomment-13376383)).

> Even if swizzling remains active, enrichScope shouldn’t invoke any UI code without dispatching to the main thread first.
>
> In the code, the window is access like this: "NSArray<UIWindow *> *appWindows = SentryDependencyContainer.sharedInstance.application.windows;"
> 
> Inside, the code is dispatched to the main thread with:
> 
> [SentryDependencyContainer.sharedInstance.dispatchQueueWrapper
>    dispatchSyncOnMainQueue:^{
>        // …
>    }];
> Shouldn’t there also be a dispatch for the line "UIScreen *appScreen = appWindows.firstObject.screen;" to ensure that this access also happens on the main thread?

I believe this is a correct assumption and could solve the access of the UI on the main thread.
We should most likely move the entire access to the screen size to the `SentryUIApplication` so we are not accessing any UIKit objects from another thread. I created #5424 to fix this.

> I just want to make sure that none of the app’s network requests can fail because of this.
> I don’t want to disable the swizzling itself, since it’s very helpful.
> Do you have any other ideas what I could do here? Or am I perhaps doing something wrong?

In general it is best practice to not close the SDK during app runtime and only bootstrap if you really want it (i.e. after a user has given consent), because the SDK is using quite a lot of static state (especially when wrapping crash handler and swizzling).

I just noticed myself that [our documentation](https://docs.sentry.io/platforms/apple/guides/ios/configuration/draining/) does not clearly state this, so we'll need to clarify here from our side.

I also recently found evidence that the `SentrySDK.close` is not correctly cleaning up the entire memory, which is tracked in #5393.
